From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nikolajs Agafonovs <nikolajs.agafonovs@gmail.com>
Date: Wed, 21 Oct 2026 16:20:08 +0300
Subject: [PATCH] tccgen: implement GNU __auto_type

parse_btype() marks the declaration (int is a placeholder) and
decl_initializer_alloc() saves the initializer, parses it once without
code to get its type (storage bits dropped, arrays and functions
decayed, a bit-field reduced to its integer type) and then parses it
for real.  An __auto_type declaration without an initializer or with a
braced initializer is an error.
---
 tcc.h    |  1 +
 tccgen.c | 28 ++++++++++++++++++++++++++++
 tcctok.h |  1 +
 3 files changed, 30 insertions(+)

diff --git a/tcc.h b/tcc.h
index 864a4e3..6052478 100644
--- a/tcc.h
+++ b/tcc.h
@@ -676,6 +676,7 @@ typedef struct AttributeDef {
     int asm_label; /* associated asm label */
     char attr_mode; /* __attribute__((__mode__(...))) */
     char new_section; /* section is new */
+    char auto_type; /* GNU '__auto_type': the type is that of the initializer */
 } AttributeDef;
 
 /* inline functions */
diff --git a/tccgen.c b/tccgen.c
index 0a122b4..337771f 100644
--- a/tccgen.c
+++ b/tccgen.c
@@ -4806,6 +4806,12 @@ static int parse_btype(CType *type, AttributeDef *ad, int ignore_label)
         case TOK_INT:
             u = VT_INT;
             goto basic_type;
+        case TOK___AUTO_TYPE:
+            /* GNU C: the type is taken from the initializer, see
+               decl_initializer_alloc(); 'int' is only a placeholder */
+            ad->auto_type = 1;
+            u = VT_INT;
+            goto basic_type;
         case TOK_ALIGNAS:
             { int n;
               AttributeDef ad1;
@@ -8310,6 +8316,28 @@ static void decl_initializer_alloc(CType *type, AttributeDef *ad, int r,
 #endif
     init_params p = {0};
 
+    if (ad->auto_type) {
+        /* GNU '__auto_type': deduce the type from the initializer, which
+           is parsed a first time without generating code */
+        if (!has_init || has_init == 2 || tok == '{')
+            tcc_error("'__auto_type' requires an expression as initializer");
+        skip_or_save_block(&init_str);
+        unget_tok(0);
+        begin_macro(init_str, 1);
+        next();
+        expr_type(type, expr_eq);
+        /* as with typeof, except that arrays and functions decay and
+           a bit-field yields its underlying integer type */
+        type->t &= ~(VT_STORAGE | VT_ARRAY | VT_VLA);
+        if (type->t & VT_BITFIELD)
+            type->t &= ~(VT_BITFIELD | (0xfffU << VT_STRUCT_SHIFT));
+        if ((type->t & VT_BTYPE) == VT_FUNC)
+            mk_pointer(type);
+        /* prepare the real initializer parsing */
+        macro_ptr = init_str->str;
+        next();
+    }
+
     if (scope == VT_CONST) {
         /* see if a global symbol was already defined */
         sym = sym_find(v);
diff --git a/tcctok.h b/tcctok.h
index d8e7b44..51676ee 100644
--- a/tcctok.h
+++ b/tcctok.h
@@ -67,6 +67,7 @@
      DEF(TOK_TYPEOF1, "typeof")
      DEF(TOK_TYPEOF2, "__typeof")
      DEF(TOK_TYPEOF3, "__typeof__")
+     DEF(TOK___AUTO_TYPE, "__auto_type")
      DEF(TOK_LABEL, "__label__")
 
 /*********************************************************************/
